File: leetcode/409.最长回文串.cpp
/*
 * @lc app=leetcode.cn id=409 lang=cpp
 *
 * [409] 最长回文串
 */
#include <iostream>
#include <utility>
#include <string>
#include <string.h>
#include <vector>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>

using namespace std;

// @lc code=start
int ah[300];
class Solution {
public:
    int longestPalindrome(string s) {
        fill(begin(ah), end(ah), 0);
        for (auto c: s) {
            ++ah[c];
        }
        int ans = 0;
        bool exist = false;
        for (int i = 0; i < 300; ++i) {
            ans += ((ah[i] >> 1) << 1);
            if (!exist && ah[i] & 0x1) {
                exist = true;
            }
        }
        if (exist) {
            ++ans;
        }
        return ans;
    }
};
// @lc code=end

int main() {
    string s = "abccccdd";
    cout << Solution{}.longestPalindrome(s);
}

File: python/409.最长回文串.py
#
# @lc app=leetcode.cn id=409 lang=python3
#
# [409] 最长回文串
#

# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> int:
        dt = {}
        for v in s:
            if v in dt:
                dt[v] += 1
            else:
                dt[v] = 1
        ans = 0
        extra = 0
        for k, v in dt.items():
            if v % 2 != 0:
                extra = 1
                v -= 1
            ans += v
        return ans + extra
# @lc code=end

